source/extensions/filters/http/istio_stats/istio_stats.cc

```diff
@@ -1132,19 +1136,32 @@ class IstioStatsFilter : public Http::PassThroughFilter,
       switch (config_->reporter()) {
       case Reporter::ServerGateway: {
         std::optional<Istio::Common::WorkloadMetadataObject> endpoint_peer;
+        // INFO: "wasm.upstream_peer" is the filter state returned
         const auto* endpoint_object = peerInfo(Reporter::ClientSidecar, filter_state);
         if (endpoint_object) {
           endpoint_peer.emplace(Istio::Common::convertFlatNodeToWorkloadMetadata(*endpoint_object));
         }
         tags_.push_back(
             {context_.destination_workload_,
              endpoint_peer ? pool_.add(endpoint_peer->workload_name_) : context_.unknown_});
-        tags_.push_back({context_.destination_workload_namespace_, context_.namespace_});
-        tags_.push_back({context_.destination_principal_,
-                         !local_san.empty() ? pool_.add(local_san) : context_.unknown_});
+        tags_.push_back({context_.destination_workload_namespace_,
+                         endpoint_peer && !endpoint_peer->namespace_name_.empty()
+                             ? pool_.add(endpoint_peer->namespace_name_)
+                             : context_.unknown_});
+
+        const auto* gateway_peer_principal  = info.filterstate().getDataReadOnly<Envoy::Extensions::Filters::Common::Expr::CelState>("wasm.upstream_peer").getDataReadOnly<Router::StringAccessor>("io.istio.peer_principal");
```

Comment:
Will clean up the code once I know it works. But the thought process here is to get the `wasm.upstream_peer` filtered state, and then get the  `io.istio.peer_principal` data from that. 

The `endpoint_peer` is stripped to only have data relevant to the [workload](https://github.com/istio/proxy/blob/f0c8cb2a2aba6b03686d2206e2266ab3eb8c8fa7/extensions/common/metadata_object.cc#L238).